src/effects/effectprocessor.h

```diff
@@ -6,100 +7,208 @@
+                          "EffectState for input" << inputHandle
+                       << "and output" << outputHandle
+                       << "EffectState should have been preallocated in the main thread.";
+            pState = createState(bufferParameters);
+            m_channelStateMatrix[inputHandle][outputHandle] = pState;
         }
-        return holder.state;
+        processChannel(inputHandle, pState, pInput, pOutput, bufferParameters,
+                       enableState, groupFeatures);
     }
 
-    ChannelHandleMap<ChannelStateHolder> m_channelState;
+   void initialize(const QSet<ChannelHandleAndGroup>& activeInputChannels,
+            EffectsManager* pEffectsManager,
+            const mixxx::AudioParameters& bufferParameters) final {
+        for (const ChannelHandleAndGroup& inputChannel : activeInputChannels) {
+            //qDebug() << this << "EffectProcessorImpl::initialize allocating EffectStates for input" << inputChannel;
+            ChannelHandleMap<EffectSpecificState*> outputChannelMap;
+            for (const ChannelHandleAndGroup& outputChannel :
+                    pEffectsManager->registeredOutputChannels()) {
+                outputChannelMap.insert(outputChannel.handle(),
+                        createState(bufferParameters));
+                //qDebug() << this << "EffectProcessorImpl::initialize registering output" << outputChannel;
+            }
+            m_channelStateMatrix.insert(inputChannel.handle(), outputChannelMap);
+        }
+        m_pEffectsManager = pEffectsManager;
+        DEBUG_ASSERT(m_pEffectsManager != nullptr);
+    };
+
+    EffectSpecificState* createState(const mixxx::AudioParameters& bufferParameters) final {
+        return new EffectSpecificState(bufferParameters);
+    };
+
+    bool loadStatesForInputChannel(const ChannelHandle& inputChannel,
+              const EffectStatesPointer pStates) final {
+          // Can't directly cast a ChannelHandleMap from containing the base
+          // EffectState* type to EffectSpecificState* type, so iterate through
+          // the ChannelHandleMap to build a new ChannelHandleMap with
+          // dynamic_cast'ed states.
+          ChannelHandleMap<EffectSpecificState*> newMap;
+          //qDebug() << "EffectProcessorImpl::loadStatesForInputChannel" << this << "input"
+          //         << inputChannel;
+          for (const ChannelHandleAndGroup& outputChannel :
+                  m_pEffectsManager->registeredOutputChannels()) {
+              //qDebug() << "EffectProcessorImpl::loadStatesForInputChannel" << this << "output" << outputChannel;
+              auto pState = dynamic_cast<EffectSpecificState*>(
+                        pStates->at(outputChannel.handle()));
+              VERIFY_OR_DEBUG_ASSERT(pState != nullptr) {
+                    return false;
+              }
+              newMap.insert(outputChannel.handle(), pState);
+          }
+          m_channelStateMatrix.insert(inputChannel, newMap);
```

Comment:
please VERIFY_OR_DEBUG_ASSERT if the array index is empty before. If not,we need to delete the old one.

Reply:
Or we need to delete the new one ... 
Depends on the decision in EngineEffectChain::enableForInputChannel